packages/contracts/contracts/L1/ZKVerifier.sol

```diff
@@ -715,14 +1190,21 @@ contract ZKVerifier {
 
     function verify(
         uint256[] calldata proof,
-        uint256[] calldata target_circuit_final_pair
+        uint256[] calldata target_circuit_final_pair,
+        bytes32 publicInputHash
     ) public view returns (bool) {
-        uint256[4] memory instances;
+        uint256[6] memory instances;
         instances[0] = target_circuit_final_pair[0] & ((1 << 136) - 1);
-        instances[1] = (target_circuit_final_pair[0] >> 136) + ((target_circuit_final_pair[1] & 1) << 136);
+        instances[1] =
+            (target_circuit_final_pair[0] >> 136) +
+            ((target_circuit_final_pair[1] & 1) << 136);
         instances[2] = target_circuit_final_pair[2] & ((1 << 136) - 1);
-        instances[3] = (target_circuit_final_pair[2] >> 136) + ((target_circuit_final_pair[3] & 1) << 136);
+        instances[3] =
+            (target_circuit_final_pair[2] >> 136) +
+            ((target_circuit_final_pair[3] & 1) << 136);
 
+        instances[4] = uint256(bytes32(bytes16(publicInputHash)) >> (8 * 16));
+        instances[5] = uint256(bytes32(bytes16(publicInputHash << (8 * 16))) >> (8 * 16));
```

Comment:
We can save gas with this code below! If i tested at local, it saves like 30 gas.

```suggestion
        instances[4] = uint256(publicInputHash) >> (8 * 16);
        instances[5] = uint256(publicInputHash) & uint256(2**128 - 1);
```